src/main.py

```diff
@@ -1,8 +1,94 @@
-from fastapi import FastAPI
+from fastapi import FastAPI, HTTPException
+from typing import List, Optional
+from src.schemas import Project, ProjectWithNoID
 
-app = FastAPI()
+project_CRUD_srvice = FastAPI()
 
 
-@app.get("/")
-async def root() -> dict[str, str]:
-    return {"message": "Hello World"}
+class ProjectServiceInterface:
+    def create_project(self, project: ProjectWithNoID) -> Optional[Project]:
+        pass
+
+    def get_projects(self) -> Optional[List[Project]]:
+        pass
+
+    def get_project_details(self, proj_id: int) -> Optional[Project]:
+        pass
+
+    def update_project_details(self, project_id: int, name: str, description: str) -> Optional[Project]:
+        pass
+
+    def delete_project(self, project_id: int) -> Optional[Project]:
+        pass
+
+
+class ProjectService(ProjectServiceInterface):
+    def __init__(self) -> None:
+        self.projects: List[Project] = []
+
+    def create_project(self, project: ProjectWithNoID) -> Project:
+        max_id = max([p.project_id for p in self.projects], default=0)
+        project_id = max_id + 1
+        project_data = project.model_dump()
+        new_project = Project(project_id=project_id, **project_data)
+
+        self.projects.append(new_project)
+        return new_project
+
+    def get_projects(self) -> List[Project]:
+        return self.projects
+
+    def get_project_details(self, proj_id: int) -> Optional[Project]:
+        project = next((p for p in self.projects if p.project_id == proj_id), None)
+        return project
+
+    def update_project_details(self, project_id: int, name: str, description: str | None) -> Optional[Project]:
+        project = next((p for p in self.projects if p.project_id == project_id), None)
+        if project is None:
+            return None
+        project.description = description
+        project.name = name
+        return project
+
+    def delete_project(self, project_id: int) -> Optional[Project]:
+        project = next((p for p in self.projects if p.project_id == project_id), None)
+        if project is None:
+            return None
+        self.projects.remove(project)
```

Comment:
We go through list two times - first to find the object and second to remove it. Let's make it more optimal